src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1201,11 +1201,33 @@ MarkRanges MergeTreeDataSelectExecutor::markRangesFromPKRange(
             * If fits, split it into smaller ones and put them on the stack. If not, discard it.
             * If the segment is already of one mark length, add it to response and discard it.
             */
-        std::vector<MarkRange> ranges_stack{ {0, marks_count} };
+        std::vector<MarkRange> ranges_stack = { {0, marks_count} };
+
+        std::function<void(size_t, size_t, FieldRef &)> create_field_ref;
+        /// If there are no monotonic functions, there is no need to save block reference.
+        /// Passing explicit field to FieldRef allows to optimize ranges and shows better performance.
```

Comment:
How does this help with performance when no monotonic functions are used?

Reply:
It avoids extra call of IColumn::operator[], which makes a copy. In first implementation, it was called every time, when we want to get field and this comment was more relevant.

Reply:
Ok, if that's the case perhaps you could add a dedicated perf test for that?